### DISCUSSION TITLE
Automated Test Suite

### DISCUSSION THREAD

**Discussion by vHanda** [jotaen/xit#6] [Tooling]
It would be awesome if I could write my own parser and then have some way to verify my parser's implementation. Basically an automated test independent of a programming language.

One easy way would be to have some official AST, and have a robust data set for '.xit' file and its AST. One easy AST could be protobufs. The [Beancount](https://github.com/Beancount/beancount) project has one for its format. It's lovely.

**Comment by jotaen** [comment]
Cool ,thanks for bringing this up, this idea actually never occurred to me. I’d need to look into that more closely.

As an aside, for my Sublime Text Package, I use a [manual test file](https://github.com/jotaen/xit-sublime/blob/main/resources/test.xit) at the moment, to at least spot check immediate issues. But that’s far from a comprehensive (and automated) test procedure.

On the other hand, a neatly composed example file could also be useful to provide. Because for example in Sublime Text I’m not sure you’d even be able to retrieve the AST of your custom syntax. (But maybe it is, I’d also need to look that up.)

**Reply by vHanda** [reply]
Another automated way might be to implement a strict code highlighter. This way any other implementation can generate a similar one, and compare images / html-subset.

**Reply by jotaen** [reply]
Just for reference, by the way, here is a screenshot of what the [manual `test.xit`](https://github.com/jotaen/xit-sublime/blob/main/resources/test.xit) file looks like with the [Sublime Text Package](https://packagecontrol.io/packages/xit!).

The Sublime Package is also not perfect, for example it erroneously highlights all occurrences of due dates within a description, whereas technically [only the first one should be recognised](https://github.com/jotaen/xit/blob/main/Specification.md#due-date).

<img width="371" alt="syntax-check" src="https://user-images.githubusercontent.com/3618384/161397282-e2b4d3eb-9b38-46a6-9b22-4ca5801450e7.png">
